solr/core/src/java/org/apache/solr/update/AddUpdateCommand.java

```diff
@@ -262,6 +263,11 @@ private void flattenAnonymous(List<SolrInputDocument> unwrappedDocs, SolrInputDo
     flattenAnonymous(unwrappedDocs, currentDoc, false);
   }
 
+  public String getRouteFieldVal() {
```

Comment:
I'm skeptical of this method.  It's name seems innocent enough looking at the code here.  But then also consider some collections have a "router.field" and this method is named in such a way that one would think this method returns that field's value... yet it does not.  Some callers put this into a variable named "id" or similar.  Given that, I propose you remove it but incorporate the logic into getHashableId which seems the proper place for it.  It _is_ the hashable Id... the hashable ID of a nested doc is it's root.

But... I do also wonder if we need this at all.  Somewhere Solr already has code that looks at \_route\_ and acts on that if present.  Perhaps the code path for an atomic update isn't doing this yet but should do it?  Then we wouldn't need this change to AddUpdateCommand.

Reply:
After reading through the code it seems like _route_ is currently only used in delete commands and queries.
This can be seen in DistributedUpdateProceesor#doDeleteByQuery
`ModifiableSolrParams outParams = new ModifiableSolrParams(filterParams(req.getParams()));
      outParams.set(DISTRIB_UPDATE_PARAM, DistribPhase.TOLEADER.toString());
      outParams.set(DISTRIB_FROM, ZkCoreNodeProps.getCoreUrl(
          zkController.getBaseUrl(), req.getCore().getName()));

      SolrParams params = req.getParams();
      String route = params.get(ShardParams._ROUTE_);
      Collection<Slice> slices = coll.getRouter().getSearchSlices(route, params, coll);

      List<Node> leaders =  new ArrayList<>(slices.size());
      for (Slice slice : slices) {
        String sliceName = slice.getName();
        Replica leader;
        try {
          leader = zkController.getZkStateReader().getLeaderRetry(collection, sliceName);
        } catch (InterruptedException e) {
          throw new SolrException(ErrorCode.SERVICE_UNAVAILABLE, "Exception finding leader for shard " + sliceName, e);
        }`
Perhaps this should be moved to addUpdateCommand(perhaps even UpdateCommand)
or to a method that will be accessible and visible so we do not have this confusion in the future?

Reply:
What code duplication?

I think we need to standardize/harmonize route a bit.  Notice UpdateCommand has setRoute & getRoute.  Lets initialize route in the constructor here by the presence of \_route\_ in the params.  Then lets not look for _route_ in params later since we know we can get it here.  Then I think `org.apache.solr.update.processor.DistributedUpdateProcessor#setupRequest(java.lang.String, org.apache.solr.common.SolrInputDocument)` can be removed and instead insist everyone call the overloaded version that takes a route, and each caller looks up the route from the command.  It's not clear to me if "null" actually should be passed to route in any circumstance.